lib/bq_stream/archive.rb

```diff
@@ -0,0 +1,239 @@
+          @oldest_attr_recs = OldestRecord.where(table_name: table)
+
+          assign_earliest_record_id(table)
+          assign_back_date_id(table.constantize, back_date)
+          
+          # Continue to work on one table until all records to back date are sent to BigQuery
+          until OldestRecord.where(table_name: table, archived: false).empty?
+            gather_records_for_buffer(table)
+
+            # Only write to BigQuery if we have rows set in the buffer
+            write_buffer_to_bq(table) unless @buffer.empty?
+          end
+        end
+      end
+    end
+
+    # Set id of the earliest record in BigQuery or first record (desc) from the database
+    def assign_earliest_record_id(table)
+      # See if there are any records for given table in BigQuery
+      record_id_query = @bq_writer.query('SELECT table_name, attr, min(record_id) as earliest_record_id '\
+                                          "FROM [#{project_id}:#{@bq_writer.dataset}.#{bq_table_name}] "\
+                                          "WHERE table_name = '#{table}' AND attr = 'id' "\
+                                          'GROUP BY table_name, attr')
+    
+      # Set earliest record id based on rows in BigQuery or the latest record in the database
+      @earliest_record_id = 
+        if record_id_query['rows']
+          record_id_query['rows'].first['f'].last['v'].to_i
+        else
+          table.constantize.unscoped.order(id: :desc).limit(1).first.try(:id)
+        end
+      
+      log(:info, "#{Time.now}: ***** earliest_record_id: #{@earliest_record_id} *****")
+    end
+
+    # Set id of the first record from back date from the database
+    def assign_back_date_id(table_class, back_date)
+      @back_date_id =
+        # See if the given table has a created_at column
+        if table_class.column_names.include?('created_at')
+          table_class.unscoped.order(id: :asc).where('created_at >= ?', back_date).limit(1).first.try(:id)
+        else
+          # Grab very first id if there is no created_at column
+          table_class.unscoped.order(id: :asc).limit(1).first.try(:id)
+        end
+      log(:info, "#{Time.now}: ***** back_date_record_id: #{@back_date_id} *****")
+    end
+
+    def gather_records_for_buffer(table)
+      # Grab records between earliest written id and back date idea
+      # limited to the number of records we can grab, keeping under 10_000 rows
+      @next_batch = table.constantize.unscoped.order(id: :desc).where('id > ? AND id <= ?', @back_date_id, @earliest_record_id).limit(10_000 / (@oldest_attr_recs.count.zero? ? 1 : @oldest_attr_recs.count)) rescue []
+      log(:info, "#{Time.now}: ***** Next Batch Count for #{table}: #{@next_batch.count} *****")
+
+      if @next_batch.empty?
+        # If there are no records in range mark the table's attributes as archived
+        OldestRecord.where(table_name: table).each { |row| row.update(archived: true) }
```

Comment:
Instead of instantiating every row, using `update_all` is significantly faster

```ruby
OldestRecord.where(table_name: table).update_all(archived: true)
```